src/gencost/waterfall.py

```diff
@@ -1708,65 +1841,194 @@ def get_cems_by_x(self, subplant_id_col, xwalk=None, merge_only=False):
-        logger.warning(
-            "CEMS %s: Unique plants that did not have matches in both "
-            "CEMS and the xwalk so will be dropped:\n %s \n",
-            {"pf_subplant_id": "prime", "subplant_id": "subplant"}[subplant_id_col],
-            test.squeeze().to_dict(),
-        )
-        if merge_only:
-            return merged
-        aggs = {
-            "generator_starts": "sum",
-            "fuel_starts": "sum",
-            "gross_generation_mwh": "sum",
-            "heat_in_mmbtu": "sum",
-            "co2_tons": "sum",
-        }
-        return (
-            merged.query("_merge == 'both'")
-            .drop(columns=["_merge"])
-            .groupby(["plant_id_eia", subplant_id_col, "report_date"])
-            .agg(aggs | {"camd_capacity_mw": "sum"})
-            .reset_index()
-            .groupby(
-                [
-                    "plant_id_eia",
-                    subplant_id_col,
-                    pd.Grouper(key="report_date", freq="YS"),
-                ],
-                dropna=False,
+                .merge(
+                    xwalk[
+                        [
+                            "plant_id_epa",
+                            "plant_id_eia",
+                            "emissions_unit_id_epa",
+                            subplant_id_col,
+                            "capacity_xwalk",
+                            "subplant_id",
+                        ]
+                    ]
+                    .dropna(axis=0, subset=["plant_id_epa", "emissions_unit_id_epa"])
+                    .drop_duplicates(),
+                    left_on=[
+                        "plant_id_epa",
+                        "emissions_unit_id_epa",
+                        "camd_capacity_mw",
+                    ],
+                    right_on=[
+                        "plant_id_epa",
+                        "emissions_unit_id_epa",
+                        "capacity_xwalk",
+                    ],
+                    how="outer",
+                    validate="m:m",
+                    indicator=True,
+                )
+                .query('_merge == "both"')
+                .groupby(
+                    [
+                        "plant_id_eia",
+                        "subplant_id",
+                        pd.Grouper(key="report_date", freq="YS"),
+                    ]
+                )
+                .agg(
+                    {
+                        "gross_generation_mwh": "sum",
+                        "generator_starts": "sum",
+                        "fuel_starts": "sum",
+                    }
+                )
+                .reset_index()
+            )
+
+            merged_with_gf_frac = (
+                self.pudl_tabl.gen_fuel_by_generator_energy_source_eia923()
+                .pipe(add_fuel_group)
+                .groupby(
+                    [
+                        "plant_id_eia",
+                        "generator_id",
+                        "fuel_group",
```

Comment:
I wonder if we could avoid dealing with the by fuel allocation step here. In other words, here in the cems function, we would worry only about allocating CEMS data to generators. Then, in `get_exa_by_generator` we could allocate gross generation at the generator level to fuel columns, maybe using `allocate_col_by`. I even think `get_gf923_by_generator` is more or less already providing the needed data for that already.

I think that obviates the `pivot_table` here too...